In this notebook, you'll see how to connect to a Postgres database using the sqlalchemy library.

For this notebook, you'll need both the `sqlalchemy` and `psycopg2` libraries installed.

In [1]:
from sqlalchemy import create_engine

First, we need to create a connection string. The format is

 ```<dialect(+driver)>://<username>:<password>@<hostname>:<port>/<database>```

To connect to the Lahman baseball database, you can use the following connection string.

In [2]:
database_name = 'scooters'    # Fill this in with your scooter database name

connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

Now, we need to create an engine and use it to connect.

In [3]:
engine = create_engine(connection_string)

Now, we can create our query and pass it into the `.query()` method.

In [4]:
# Look at difference in run time for this:
query = '''
SELECT latitude
FROM scooters;
'''

result = engine.execute(query)

In [5]:
# Vs this:
query = '''
SELECT COUNT(latitude)
FROM scooters;
'''

result = engine.execute(query)

You can then fetch the results as tuples using either `fetchone` or `fetchall`:

In [6]:
result.fetchone()

(73414043,)

On the other hand, sqlalchemy plays nicely with pandas.

In [7]:
import pandas as pd

In [8]:
lat = pd.read_sql(query, con = engine)
lat.head()

count
0  73414043

Are there any null values in any columns in either table?

In [9]:
query = '''

SELECT COUNT(*)
FROM scooters
WHERE pubdatetime IS NULL
'''

In [10]:
pubdatetimenulls = pd.read_sql(query, con=engine)
print(pubdatetimenulls)

   count
0      0


In [11]:
query = '''

SELECT COUNT(*)
FROM scooters
WHERE chargelevel IS NULL

'''

In [12]:
chargelevel_nulls = pd.read_sql(query, con=engine)
print(chargelevel_nulls)

   count
0    770


In [13]:
query= '''
SELECT SUM(CASE WHEN chargelevel is null THEN 1 ELSE 0 END) 
AS charge_nulls 
FROM scooters
'''

In [14]:
charge_nulls = pd.read_sql(query, con=engine)
print(charge_nulls)

   charge_nulls
0           770


In [15]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE enddate IS NULL

'''


In [16]:
enddate_nulls = pd.read_sql(query, con=engine)
print(enddate_nulls)

   count
0      0


In [17]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE endtime IS NULL

'''

In [18]:
endtime_nulls = pd.read_sql(query, con=engine)
print(endtime_nulls)

   count
0      0


In [19]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE startlatitude IS NULL

'''

In [20]:
startlatitude_nulls = pd.read_sql(query, con=engine)
print(startlatitude_nulls)

   count
0      0


In [21]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE startlongitude IS NULL

'''

In [22]:
startlongitude_nulls = pd.read_sql(query, con=engine)
print(startlongitude_nulls)

   count
0      0


In [23]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE endlatitude IS NULL

'''

In [24]:
endlatitude_nulls = pd.read_sql(query, con=engine)
print(endlatitude_nulls)

   count
0      0


In [25]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE endlongitude IS NULL

'''

In [26]:
endlongitude_nulls = pd.read_sql(query, con=engine)
print(endlongitude_nulls)

   count
0      0


In [27]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE triproute IS NULL

'''

In [28]:
triproute_nulls = pd.read_sql(query, con=engine)
print(triproute_nulls)

   count
0      0


In [29]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE create_dt IS NULL

'''

In [30]:
create_dt_nulls = pd.read_sql(query, con=engine)
print(create_dt_nulls)

   count
0      0


In [31]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE sumdid IS NULL

'''

In [32]:
sumdid_nulls = pd.read_sql(query, con=engine)
print(sumdid_nulls)

   count
0      0


In [33]:
query = '''

SELECT COUNT(*)
FROM trips
WHERE tripdistance IS NULL

'''

In [34]:
tripdistance_nulls = pd.read_sql(query, con=engine)
print(tripdistance_nulls)

   count
0      0


Answer: 770 nulls on scooters chargelevel


What date range is represented in each of the date columns? Investigate any values that seem odd.

In [35]:
query = '''SELECT MIN(pubdatetime)
FROM scooters
'''

In [36]:
min_scooter_pubdatetime = pd.read_sql(query, con=engine)
print(min_scooter_pubdatetime)

                      min
0 2019-05-01 00:01:41.247


In [37]:
query = '''SELECT MAX(pubdatetime)
FROM scooters
'''

In [38]:
max_scooter_pubdatetime = pd.read_sql(query, con=engine)
print(max_scooter_pubdatetime)

                  max
0 2019-07-31 23:59:57


In [39]:
query = '''SELECT MIN(pubtimestamp)
FROM trips
'''

In [40]:
min_trips_pubtimestamp = pd.read_sql(query, con=engine)
print(min_trips_pubtimestamp)

                      min
0 2019-05-01 00:00:55.423


In [41]:
query = '''SELECT MAX(pubtimestamp)
FROM trips
'''

In [42]:
max_trips_pubtimestamp = pd.read_sql(query, con=engine)
print(max_trips_pubtimestamp)

                  max
0 2019-08-01 07:04:00


In [43]:
query = '''SELECT DISTINCT sumdgroup
FROM scooters
'''


In [44]:
distinct_sumdgroup = pd.read_sql(query, con=engine)
print(distinct_sumdgroup)

  sumdgroup
0   bicycle
1   scooter
2   Scooter


What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [45]:
query = '''SELECT MIN(startlatitude)
FROM trips
'''

In [46]:
min_start_latitude = pd.read_sql(query, con=engine)
print(min_start_latitude)

       min
0  35.8532


In [47]:
query = '''SELECT MAX(startlatitude)
FROM trips
'''

In [48]:
max_start_latitude = pd.read_sql(query, con=engine)
print(max_start_latitude)

         max
0  36.300029


In [49]:
query = '''SELECT MIN(startlongitude)
FROM trips
'''

In [50]:
min_start_longitude = pd.read_sql(query, con=engine)
print(min_start_longitude)

         min
0 -86.918008


In [51]:
query = '''SELECT MAX(startlongitude)
FROM trips
'''

In [52]:
max_start_longitude = pd.read_sql(query, con=engine)
print(max_start_longitude)

       max
0 -86.3662


In [53]:
query = '''SELECT MAX(tripduration)
FROM trips
'''

In [54]:
max_trip_duration = pd.read_sql(query, con=engine)
print(max_trip_duration)

        max
0  512619.0


In [55]:
query = '''SELECT MIN(tripduration)
FROM trips
'''

In [56]:
min_trip_duration = pd.read_sql(query, con=engine)
print(min_trip_duration)

         min
0 -19.358267


1. During this period, seven companies offered scooters. How many scooters did each company have in this time frame? Did the number for each company change over time? Did scooter usage vary by company?

In [59]:
query = '''SELECT DISTINCT companyname
FROM scooters'''

In [60]:
scooter_companies = pd.read_sql(query, con=engine)
print(scooter_companies)

  companyname
0        Bird
1        Bolt
2      Gotcha
3        Jump
4        Lime
5        Lyft
6        Spin


In [61]:
import pandas as pd

In [62]:
import numpy as np

In [75]:
import matplotlib as plt

In [64]:
query = '''SELECT COUNT(sumdid)
FROM scooters
WHERE companyname = 'Bird'
'''

In [65]:
bird_count = pd.read_sql(query, con=engine)
print(bird_count)

      count
0  12251590


In [68]:
query = '''SELECT COUNT(sumdid)
FROM scooters
WHERE companyname = 'Bolt'
'''

In [69]:
bolt_count = pd.read_sql(query, con=engine)
print(bolt_count)

     count
0  3477198


In [71]:
query = '''
SELECT DISTINCT(sumdid),
companyname
FROM scooters;
'''

In [73]:
company_scooters = pd.read_sql(query, con = engine)
company_scooters.head()

sumdid companyname
0  Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt
1  Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d        Bolt
2  Powered-03be23ca-d43b-222f-be54-e44b5b4690df        Bolt
3  Powered-046201fb-6532-1f37-6334-3612fb1e61f7        Bolt
4  Powered-0479bb84-afbd-0426-f1c4-df628542a88c        Bolt

In [74]:
df2 = company_scooters.groupby(['companyname'])['companyname'].count()
print(df2)

companyname
Bird      3860
Bolt       360
Gotcha     224
Jump      1210
Lime      1824
Lyft      1735
Spin       805
Name: companyname, dtype: int64


In [76]:
query = '''
SELECT sumdid,
companyname, MIN(pubdatetime)
FROM scooters
GROUP BY sumdid, companyname;
'''

company_scooters_time = pd.read_sql(query, con = engine)
company_scooters_time.head()

sumdid companyname   
0  Powered-017d3133-f14a-2b83-ee4f-d777e7c5b619        Bolt  \
1  Powered-01a24436-0315-e1bb-7ce0-d081d05dff7d        Bolt   
2  Powered-03be23ca-d43b-222f-be54-e44b5b4690df        Bolt   
3  Powered-046201fb-6532-1f37-6334-3612fb1e61f7        Bolt   
4  Powered-0479bb84-afbd-0426-f1c4-df628542a88c        Bolt   

                  min  
0 2019-05-24 00:04:42  
1 2019-05-24 00:04:42  
2 2019-05-24 00:04:42  
3 2019-05-28 20:53:56  
4 2019-05-24 00:04:42

In [77]:
import seaborn as sb

In [ ]:
scooters = pd.read_csv